daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1213,90 +1201,96 @@ private[lf] object SBuiltin {
     }
   }
 
-  private[speedy] sealed abstract class SBUKeyBuiltin(templateId: TypeConName)
-      extends OnLedgerBuiltin(1) {
+  private[this] abstract class KeyOperation {
+    val templateId: TypeConName
     private[this] val typ = AstUtil.TContractId(Ast.TTyCon(templateId))
 
-    sealed trait KeyOperation {
-      // Callback from the engine returned NotFound
-      def handleKeyNotFound(machine: Machine): Boolean
-      // We saw an archive of this key.
-      def handleKeyArchived(machine: Machine, key: GlobalKey): Unit
-      def cidToSValue(cid: V.ContractId): SValue
-      def cidToSExpr(cid: V.ContractId): SExpr
-    }
-    final case object Fetch extends KeyOperation {
-      override def handleKeyNotFound(machine: Machine): Boolean =
+    final protected def importCid(cid: V.ContractId): SEImportValue =
+      SEImportValue(typ, V.ValueContractId(cid))
+    // Callback from the engine returned NotFound
+    def handleInputKeyNotFound(machine: Machine): Boolean
+    // CallBack from the engine returned a new Cid
+    def handleInputKeyFound(machine: Machine, cid: V.ContractId): Unit
+    // We already saw this key, but it was undefined or was archived
+    def handleInactiveKey(machine: Machine, gkey: GlobalKey): Unit
+    // We already saw this key and it is still active
+    def handleActiveKey(machine: Machine, cid: V.ContractId): Unit
+
+    final def handleKnownInputKey(
+        machine: Machine,
+        gkey: GlobalKey,
+        keyMapping: PartialTransaction.KeyMapping,
+    ): Unit =
+      keyMapping match {
+        case PartialTransaction.KeyActive(cid) => handleActiveKey(machine, cid)
+        case PartialTransaction.KeyInactive => handleInactiveKey(machine, gkey)
+      }
+  }
+
+  private[this] object KeyOperation {
+    final class Fetch(override val templateId: TypeConName) extends KeyOperation {
```

Comment:
why does the operation need to store the template id?

Reply:
`SEImport` used it in the case of `handleInputKeyFound`, required the type. 
The type is actually not used in the case of single contract ID, we could use a dummy (wrong !) type instead but I do not really like that.